File: src/utils/survey.js
/* 설문(Survey) API 클라이언트.
 *
 * 백엔드 라우터: server/app/routers/survey.py (prefix /api/survey).
 * 이 프로젝트의 다른 호출들과 동일하게 axios 가 아니라 fetch + 쿠키 인증
 * (credentials: "include") 을 쓴다.
 */

/* 현재 사용자에게 띄울 설문을 조회.
 *
 * @returns {Promise<{
 *   due: 'onboarding' | 'recurring' | null,
 *   response_id?: number,
 *   schema_id?: number,
 *   schema_version?: string,
 *   schema_json?: object,
 *   current_section?: string | null,
 *   answers?: object,
 *   prefilled_answers?: object,
 *   reward_points?: number,   // 완료 시 받는 포인트 — 진행 중 독려 안내에 사용
 * }>}
 */
export async function getNextSurvey() {
  const res = await fetch("/api/survey/next", { credentials: "include" });
  if (!res.ok) throw new Error(`survey/next ${res.status}`);
  return res.json();
}

/* 진행 중 설문에 부분 응답을 저장(섹션 단위 autosave).
 *
 * @param {number} responseId
 * @param {{ answers: object, currentSection?: string|null }} params
 *   answers: 변경분만 보내도 서버가 기존 위에 merge.
 * @returns {Promise<{ response_id, current_section, updated_at }>}
 */
export async function saveSurveyProgress(responseId, { answers, currentSection } = {}) {
  const res = await fetch(`/api/survey/${responseId}/progress`, {
    method: "PATCH",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      answers: answers || {},
      current_section: currentSection ?? null,
    }),
  });
  if (!res.ok) throw new Error(`survey/progress ${res.status}`);
  return res.json();
}

/* 설문 최종 제출. 백엔드가 scoring → derived_flags 저장 →
 * User.onboarded / last_survey_at 갱신 → SafetyEvent 적재 수행.
 *
 * @param {number} responseId
 * @returns {Promise<{ response_id, derived_flags, completed_at, points_awarded }>}
 *   points_awarded: 이번 제출로 새로 적립된 포인트(설문 완료 = 50).
 */
export async function submitSurvey(responseId) {
  const res = await fetch(`/api/survey/${responseId}/submit`, {
    method: "POST",
    credentials: "include",
  });
  if (!res.ok) throw new Error(`survey/submit ${res.status}`);
  return res.json();
}

File: server/app/schemas/survey.py
"""설문(Survey) API 요청/응답 스키마."""
from datetime import datetime

from pydantic import BaseModel


class SurveyNextResponse(BaseModel):
    """GET /api/survey/next 응답.

    due 가 None 이면 보여줄 설문 없음(나머지 필드는 의미 없음).
    """

    due: str | None = None  # "onboarding" | "recurring" | None
    response_id: int | None = None
    schema_id: int | None = None
    schema_version: str | None = None
    schema_json: dict | None = None
    current_section: str | None = None
    answers: dict = {}
    prefilled_answers: dict = {}
    reward_points: int = 0  # 설문 완료 시 받는 포인트 — 진행 중 독려 안내에 사용


class SurveyProgressRequest(BaseModel):
    """PATCH /api/survey/{id}/progress 요청 — 한 섹션 분량의 부분 응답.

    answers: 이번에 추가/변경된 question_id → 값 매핑. 서버는 기존 answers 에
        update 한다(변경분만 보내도 됨).
    current_section: 다음에 이어서 시작할 섹션 식별자 (예: "C").
    """

    answers: dict
    current_section: str | None = None


class SurveyProgressResponse(BaseModel):
    response_id: int
    current_section: str | None = None
    updated_at: datetime


class SurveySubmitResponse(BaseModel):
    response_id: int
    derived_flags: dict
    completed_at: datetime
    points_awarded: int = 0  # 이번 제출로 새로 적립된 포인트(설문 완료 = 50)
